A36-xds-for-servers.md

```diff
@@ -0,0 +1,316 @@
+"real" builder. The server returned from `builder.build()` would be an xDS-aware
+server delegating to a "real" server instance. The `XdsServerBuilder` would
+install an `AtomicReference&lt;ServerInterceptor>`-backed interceptor in the
+built server and pass the listening port and interceptor reference to the
+xDS-aware server when constructed. Since interceptors cannot be removed once
+installed, the builder may only be used once (`build()` should throw if called a
+second time).
+
+The xDS-aware server’s `start()` will create the `XdsClient` with the passed
+port and wait for initial configuration before delegating to the real `start()`.
+If the xDS bootstrap is missing it will throw an IOException within `start()`.
+The `XdsClient` will be shut down on `shutdownNow()`/termination. When new or
+updated configuration is received, it will create a new intercepter and update
+the interceptor reference.
+
+XdsServerBuilder will have an `xdsInitListener(XdsInitListener)`.
+`XdsInitListener` will have the one method `xdsLoadFailure(IOException)` that is
+called for failures loading the initial configuration within `start()`, since
+`start()` itself does not fail in that case. A default implementation will log
+the exception. If a user sets the listener, it will replace the default logging
+implementation.
+
+In order to allow transport-specific configuration, the `XdsServerBuilder` will
+have a `@ExperimentalApi ServerBuilder transportBuilder()` method whose return
+value can be cast to `NettyServerBuilder` for experimental API configuration.
+
+`NettyServerBuilder` will add an internal method `eagAttributes(Attributes)`. It
+will plumb those attributes to `GrpcHttp2ConnectionHandler.getEagAttributes()`
+as implemented by `NettyServerHandler`, which currently just returns
+`Attributes.EMPTY`. `XdsServerBuilder` will then specify `eagAttributes` to
+inject credential information for the `XdsServerCredential`, similar to
+client-side, although with the added step of needing to process the
+FilterChainMatch for the specific connection.
+
+
+### Go
+
+Create an `xds.Server` struct that would internally contain a `grpc.Server`. It
+would inject its own `StreamServerInterceptor` and `UnaryServerInterceptor`s
+into the `grpc.Server`. The interceptors would be controlled by an
+`atomic.Value` that would update with the configuration. When
+`Serve(net.Listener)` is called it creates an XdsClient and observes the
+`Addr()` to create a watch.  Before `Serve(net.Listener)` returns, it will shut
+down the XdsClient. Note that configuration is nominally per-address, but there
+is only one set of interceptors, so the interceptors will need to look up the
+per-address configuration each RPC.
+
+TODO: It looks like there is a lifetime issue for the XdsClient, as RPCs can
+continue after the listener is closed and Serve() returns.
+
+Because service registration is done on a method in the generated code, and
```

Comment:
This is no longer true. We have added a `ServiceRegistrar` interface to the `grpc` package which contains a single method `RegisterService`, and this is implemented by both `grpc.Server` and `xds.GRPCServer`. The protoc-gen-go has also been updated to emit code where the registration function accepts this interface instead of a concrete type. So, users should be able to directly pass `xds.GRPCServer` to these registration functions, and the underlying `grpc.Server` is no longer exported.

https://github.com/grpc/grpc-go/blob/8736dcd05b735913226c0c8bf787db48a13a7129/server.go#L558